# Clean Data for National Presidential Party Wins
This notebook takes in a CSV of data from a Kaggle data set and cleans it for export to CSV and use in Tableau
**Found some errors in the data so using a different source**

In [35]:
import pandas as pd

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

### Import Kaggle Data
Data source is https://www.kaggle.com/harshitagpt/us-presidents and was created by Harshita Gupta

In [36]:
input_file = 'raw_data/us_presidents.csv'
df = pd.read_csv(input_file)
df.head()

Unnamed: 0  S.No.           start            end          president  \
0           0      1  April 30, 1789  March 4, 1797  George Washington   
1           1      2   March 4, 1797  March 4, 1801         John Adams   
2           2      3   March 4, 1801  March 4, 1809   Thomas Jefferson   
3           3      4   March 4, 1809  March 4, 1817      James Madison   
4           4      5   March 4, 1817  March 4, 1825       James Monroe   

                                               prior                    party  \
0  Commander-in-Chief  of the  Continental Army  ...       Nonpartisan   [13]   
1           1st  Vice President of the United States               Federalist   
2           2nd  Vice President of the United States  Democratic-  Republican   
3  5th  United States Secretary of State   (1801–...  Democratic-  Republican   
4  7th  United States Secretary of State   (1811–...  Democratic-  Republican   

                 vice  
0          John Adams  
1    Thomas Jefferson  
2          Aaron Burr  
3      George Clinton  
4  Daniel D. Tompkins

### Remove Unnecessary Columns

In [37]:
df = df[['president', 'party', 'start', 'end']]
df.head()

president                    party           start            end
0  George Washington       Nonpartisan   [13]  April 30, 1789  March 4, 1797
1         John Adams               Federalist   March 4, 1797  March 4, 1801
2   Thomas Jefferson  Democratic-  Republican   March 4, 1801  March 4, 1809
3      James Madison  Democratic-  Republican   March 4, 1809  March 4, 1817
4       James Monroe  Democratic-  Republican   March 4, 1817  March 4, 1825

In [38]:
df

president                                        party  \
0        George Washington                           Nonpartisan   [13]   
1               John Adams                                   Federalist   
2         Thomas Jefferson                      Democratic-  Republican   
3            James Madison                      Democratic-  Republican   
4             James Monroe                      Democratic-  Republican   
5        John Quincy Adams                      Democratic-  Republican   
6           Andrew Jackson                                   Democratic   
7         Martin Van Buren                                   Democratic   
8   William Henry Harrison                                         Whig   
9               John Tyler  Whig   April 4, 1841  –  September 13, 1841   
10           James K. Polk                                   Democratic   
11          Zachary Taylor                                         Whig   
12        Millard Fillmore                                         Whig   
13         Franklin Pierce                                   Democratic   
14          James Buchanan                                   Democratic   
15         Abraham Lincoln        Republican   ( National Union )   [i]   
16          Andrew Johnson  National Union   [i]   ( Democratic )   [j]   
17        Ulysses S. Grant                                   Republican   
18     Rutherford B. Hayes                                   Republican   
19       James A. Garfield                                   Republican   
20       Chester A. Arthur                                   Republican   
21        Grover Cleveland                                   Democratic   
22       Benjamin Harrison                                   Republican   
23        Grover Cleveland                                   Democratic   
24        William McKinley                                   Republican   
25      Theodore Roosevelt                                   Republican   
26     William Howard Taft                                   Republican   
27          Woodrow Wilson                                   Democratic   
28       Warren G. Harding                                   Republican   
29         Calvin Coolidge                                   Republican   
30          Herbert Hoover                                   Republican   
31   Franklin D. Roosevelt                                   Democratic   
32         Harry S. Truman                                   Democratic   
33    Dwight D. Eisenhower                                   Republican   
34         John F. Kennedy                                   Democratic   
35       Lyndon B. Johnson                                   Democratic   
36           Richard Nixon                                   Republican   
37             Gerald Ford                                   Republican   
38            Jimmy Carter                                   Democratic   
39           Ronald Reagan                                   Republican   
40       George H. W. Bush                                   Republican   
41            Bill Clinton                                   Democratic   
42          George W. Bush                                   Republican   
43            Barack Obama                                   Democratic   
44            Donald Trump                                   Republican   

                 start                 end  
0       April 30, 1789       March 4, 1797  
1        March 4, 1797       March 4, 1801  
2        March 4, 1801       March 4, 1809  
3        March 4, 1809       March 4, 1817  
4        March 4, 1817       March 4, 1825  
5        March 4, 1825       March 4, 1829  
6        March 4, 1829       March 4, 1837  
7        March 4, 1837       March 4, 1841  
8        March 4, 1841       April 4, 1841  
9        April 4, 1841       March 4, 1845  
10       March 4, 1845       March 4, 1849  
11       March 4, 1849    

### Convert Start and End Dates to Years

In [25]:
df['start_yr'] = pd.to_datetime(df['start'], errors='coerce').dt.year
df['end_yr'] = pd.to_datetime(df['end'], errors='coerce').dt.year
df.head()

president                    party           start            end  \
0  George Washington       Nonpartisan   [13]  April 30, 1789  March 4, 1797   
1         John Adams               Federalist   March 4, 1797  March 4, 1801   
2   Thomas Jefferson  Democratic-  Republican   March 4, 1801  March 4, 1809   
3      James Madison  Democratic-  Republican   March 4, 1809  March 4, 1817   
4       James Monroe  Democratic-  Republican   March 4, 1817  March 4, 1825   

   start_yr  end_yr  
0      1789  1797.0  
1      1797  1801.0  
2      1801  1809.0  
3      1809  1817.0  
4      1817  1825.0

### Calculate Potential Election Years

In [26]:
df['election_yr'] = df['start_yr'] - 1
df.loc[df['president'] == 'George Washington', 'election_yr'] = df['start_yr']
df.head()

president                    party           start            end  \
0  George Washington       Nonpartisan   [13]  April 30, 1789  March 4, 1797   
1         John Adams               Federalist   March 4, 1797  March 4, 1801   
2   Thomas Jefferson  Democratic-  Republican   March 4, 1801  March 4, 1809   
3      James Madison  Democratic-  Republican   March 4, 1809  March 4, 1817   
4       James Monroe  Democratic-  Republican   March 4, 1817  March 4, 1825   

   start_yr  end_yr  election_yr  
0      1789  1797.0         1789  
1      1797  1801.0         1796  
2      1801  1809.0         1800  
3      1809  1817.0         1808  
4      1817  1825.0         1816

### Calculate Terms Served

In [34]:
df['election_yr'].shift(-1) - df['election_yr']

0      7.0
1      4.0
2      8.0
3      8.0
4      8.0
5      4.0
6      8.0
7      4.0
8      0.0
9      4.0
10     4.0
11     1.0
12     3.0
13     4.0
14     4.0
15     4.0
16     4.0
17     8.0
18     4.0
19     0.0
20     4.0
21     4.0
22     4.0
23     4.0
24     4.0
25     8.0
26     4.0
27     8.0
28     2.0
29     6.0
30     4.0
31    12.0
32     8.0
33     8.0
34     2.0
35     6.0
36     5.0
37     3.0
38     4.0
39     8.0
40     4.0
41     8.0
42     8.0
43     8.0
44     NaN
Name: election_yr, dtype: float64

In [28]:
(df['end_yr'] - df['start_yr']) // 4

0     2.0
1     1.0
2     2.0
3     2.0
4     2.0
5     1.0
6     2.0
7     1.0
8     0.0
9     1.0
10    1.0
11    0.0
12    0.0
13    1.0
14    1.0
15    1.0
16    1.0
17    2.0
18    1.0
19    0.0
20    1.0
21    1.0
22    1.0
23    1.0
24    1.0
25    2.0
26    1.0
27    2.0
28    0.0
29    1.0
30    1.0
31    2.0
32    2.0
33    2.0
34    0.0
35    1.0
36    1.0
37    0.0
38    1.0
39    2.0
40    1.0
41    2.0
42    2.0
43    NaN
44    NaN
dtype: float64

### Create DataFrame of Election Years and Merge Presidents Data

In [20]:
election_years = [1789] + list(range(1792, df['start_yr'].max(), 4))
election_df = pd.DataFrame(election_years, columns=['year'])
election_df = pd.merge_asof(election_df,
                            df,
                            left_on='year',
                            right_on='election_yr',
                            direction='backward')
election_df

year              president                                        party  \
0   1789      George Washington                           Nonpartisan   [13]   
1   1792      George Washington                           Nonpartisan   [13]   
2   1796             John Adams                                   Federalist   
3   1800       Thomas Jefferson                      Democratic-  Republican   
4   1804       Thomas Jefferson                      Democratic-  Republican   
5   1808          James Madison                      Democratic-  Republican   
6   1812          James Madison                      Democratic-  Republican   
7   1816           James Monroe                      Democratic-  Republican   
8   1820           James Monroe                      Democratic-  Republican   
9   1824      John Quincy Adams                      Democratic-  Republican   
10  1828         Andrew Jackson                                   Democratic   
11  1832         Andrew Jackson                                   Democratic   
12  1836       Martin Van Buren                                   Democratic   
13  1840             John Tyler  Whig   April 4, 1841  –  September 13, 1841   
14  1844          James K. Polk                                   Democratic   
15  1848         Zachary Taylor                                         Whig   
16  1852        Franklin Pierce                                   Democratic   
17  1856         James Buchanan                                   Democratic   
18  1860        Abraham Lincoln        Republican   ( National Union )   [i]   
19  1864         Andrew Johnson  National Union   [i]   ( Democratic )   [j]   
20  1868       Ulysses S. Grant                                   Republican   
21  1872       Ulysses S. Grant                                   Republican   
22  1876    Rutherford B. Hayes                                   Republican   
23  1880      Chester A. Arthur                                   Republican   
24  1884       Grover Cleveland                                   Democratic   
25  1888      Benjamin Harrison                                   Republican   
26  1892       Grover Cleveland                                   Democratic   
27  1896       William McKinley                                   Republican   
28  1900     Theodore Roosevelt                                   Republican   
29  1904     Theodore Roosevelt                                   Republican   
30  1908    William Howard Taft                                   Republican   
31  1912         Woodrow Wilson                                   Democratic   
32  1916         Woodrow Wilson                                   Democratic   
33  1920      Warren G. Harding                                   Republican   
34  1924        Calvin Coolidge                                   Republican   
35  1928         Herbert Hoover                                   Republican   
36  1932  Franklin D. Roosevelt                                   Democratic   
37  1936  Franklin D. Roosevelt                                   Democratic   
38  1940  Franklin D. Roosevelt                                   Democratic   
39  1944        Harry S. Truman                                   Democratic   
40  1948        Harry S. Truman                                   Democratic   
41  1952   Dwight D. Eisenhower                                   Republican   
42  1956   Dwight D. Eisenhower                                   Republican   
43  1960        John F. Kennedy                                   Democratic   
44  1964      Lyndon B. Johnson                                   Democratic   
45  1968          Richard Nixon                                   Republican   
46  1972          Richard Nixon                                   Republican   
47  1976           Jimmy Carter                                   Democratic   
48  1980          Ronald Reagan                                   Republican   
49  

Fixes:
1840: John Tyler should be William Harrison
1864: should be Lincoln
1880: Garfield
1900: McKinley
1944: FDR